src/GraphQL/Internal/Validation.hs

```diff
@@ -183,18 +205,42 @@ validateArguments args = mapErrors DuplicateArgument (makeMap [(name, value) | A
 -- imply that everything is valid.
 
 -- | A GraphQL selection.
-data Selection spread
-  = SelectionField (Field spread)
+data Selection' spread
+  = SelectionField (Field' spread)
   | SelectionFragmentSpread spread
   | SelectionInlineFragment (InlineFragment spread)
   deriving (Eq, Show)
 
+-- | Get all of the fields directly inside the given selection set.
+--
+-- TODO: This ignores fragments, whereas it should actually do something with
+-- them.
+--
+-- TODO: At this point, we ought to know that field names are unique. As such,
+-- we should return an ordered map of Name to Fields, rather than a bland
+-- list.
+getFields :: SelectionSet -> [Field]
+getFields ss = [field | SelectionField field <- ss]
+
 -- | A field in a selection set, which itself might have children which might
 -- have fragment spreads.
-data Field spread
-  = Field (Maybe Alias) Name Arguments Directives [Selection spread]
+data Field' spread
+  = Field' (Maybe Alias) Name Arguments Directives [Selection' spread]
   deriving (Eq, Show)
 
+instance HasName (Field' spread) where
+  getName (Field' _ name _ _ _) = name
+
+type Field = Field' FragmentSpread
```

Comment:
Yes, that's the difference. `Resolved | Unresolved` would be accurate, but not precise. More accurate would be "Field that has resolved fragment spreads". So `ResolvedFragmentSpread | UnresolvedFragmentSpread` (although note that these are different types, and you can't normally write `a | b` in Haskell). Because this module follows a convention of calling fully resolved things by their simplest name, it's `FragmentSpread | UnresolvedFragmentSpread`.

FWIW, I did write a large [comment explaining this](https://github.com/jml/graphql-api/blob/master/src/GraphQL/Internal/Validation.hs#L164-L183). If it's not helpful, please let me know how it could be more helpful.